README.md
[![PyPI version](https://badge.fury.io/py/bwsample.svg)](https://badge.fury.io/py/bwsample)
[![PyPi downloads](https://img.shields.io/pypi/dm/bwsample)](https://img.shields.io/pypi/dm/bwsample)
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03324/status.svg)](https://doi.org/10.21105/joss.03324)
[![DOI](https://zenodo.org/badge/335090754.svg)](https://zenodo.org/badge/latestdoi/335090754)
[![Join the chat at https://gitter.im/satzbeleg/community](https://badges.gitter.im/satzbeleg/community.svg)](https://gitter.im/satzbeleg/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![bwsample](https://snyk.io/advisor/python/bwsample/badge.svg)](https://snyk.io/advisor/python/bwsample)



# bwsample: Sampling and Evaluation of Best-Worst Scaling sets
Sampling algorithm for best-worst scaling (BWS) sets, extracting pairs from evaluated BWS sets, count in dictionary of keys sparse matrix, and compute scores based on it.

The package `bwsample` addresses three areas:

* [Sampling](#sampling)
* [Counting](#counting)
* [Ranking](#ranking)

## Installation
The `bwsample` [git repo](http://github.com/satzbeleg/bwsample) is available as [PyPi package](https://pypi.org/project/bwsample)

```sh
pip install "bwsample>=0.7.0"
```

## Overview
The `bwsample` can be deployed at different stages to prepare BWS example sets and post-process evaluated BWS sets.
An *Active Learning* experiment using an Web App with BWS user interface to judge sentence examples is shown in the diagram below. The `bwsample` would be implemented in a (python based) REST API. The App requests new BWS example sets, and `bwsample.sample` generates these. After the App posts the evaluation results to the API, `bwsample.count` extract new pairwise data from evaluated BWS sets. The pairwise comparision matrix can be used by `bwsample.rank` to compute  scores for a new updated training set.

![](https://raw.githubusercontent.com/satzbeleg/bwsample/main/docs/bwsample-process.png)

## Sampling
**Input Data:**
The input data `examples` for `bwsample.sample` should be a `List[anything]`.
For example, `List[Dict[ID,DATA]]` with identifiers using the key `"id"` and the associated data using the key `"data"`, e.g.

```python
examples = [
    {"id": "id1", "data": "data..."},
    {"id": "id2", "data": ["other", "data"]},
    {"id": "id3", "data": {"key", "value"}},
    {"id": "id4", "data": "lorem"},
    {"id": "id5", "data": "ipsum"},
    {"id": "id6", "data": "blind"},
    {"id": "id7", "data": "text"}
]
```

**Call the function:**
The number of items per BWS set `n_items` (`M`) must be specified, e.g. `n_items=4` if your App displays four items.
The `'overlap'` algorithm assigns every `i*(M-1)+1`-th example to two consecutive BWS sets, so that `1/(M-1)` of examples are evaluated two times.
The `'twice'` algorithm connects the remaining `(M-2)/(M-1)` non-overlapping from `'overlapping'` so that all examples occur twice.
The total number of sampled BWS sets might differ accordingly.

```python
import bwsample as bws
samples = bws.sample(examples, n_items=4, method='overlap')
```

**Output Data:**
The output has the following structure

```
[
    [{'id': 'id1', 'data': 'data...'}, {'id': 'id2', 'data': ['other', 'data']}, {'id': 'id3', 'data': {'key', 'value'}}, {'id': 'id4', 'data': 'lorem'}], 
    [{'id': 'id1', 'data': 'data...'}, {'id': 'id4', 'data': 'lorem'}, {'id': 'id5', 'data': 'ipsum'}, {'id': 'id6', 'data': 'blind'}]
]
```

**Warning**: `len(examples)` must be a multiple of `(n_items - 1)`

**References:**

- Section 5 (page 4) in: Hamster, U. A. (2021, March 9). Extracting Pairwise Comparisons Data from Best-Worst Scaling Surveys by Logical Inference. [https://doi.org/10.31219/osf.io/qkxej](https://doi.org/10.31219/osf.io/qkxej)


## Counting
**Input Data:**
The input data`evaluations` for `bwsample.count` should structured as `List[Tuple[List[ItemState], List[ItemID]]]`. 
The labelling/annotation application should produce a list of item states `List[ItemState]` with the states `BEST:1`, `WORST:2` and `NOT:0` for each item. 
And the corresponding list of IDs `List[ItemID]` for each item or resp. example.

```python
evaluations = (
    ([0, 0, 2, 1], ['id1', 'id2', 'id3', 'id4']), 
    ([0, 1, 0, 2], ['id4', 'id5', 'id6', 'id7']),
    ([1, 2, 0, 0], ['id7', 'id8', 'id9', 'id1'])
)
```

**Call the function:**

```python
import bwsample as bws
agg_dok, direct_dok, direct_detail, logical_dok, logical_detail = bws.count(evaluations)
```


**Output Data:**
The function `bwsample.count` outputs Dictionary of Keys (DOK) with the data structure `Dict[Tuple[ItemID, ItemID], int]`, e.g. `agg_dok`, `direct_dok`, `direct_detail["bw"]`, etc., what contain variants which pairs where counted:

- `agg_dok`
    - `direct_dok`
        - `direct_detail["bw"]` -- `BEST>WORST`
        - `direct_detail["bn"]` -- `BEST>NOT`
        - `direct_detail["nw"]` -- `NOT>WORST`
    - `logical_dok`
        - `logical_detail["nn"]` -- `D>E>F vs X>E>Z`
        - `logical_detail["nb"]` -- `D>E>F vs E>Y>Z`
        - `logical_detail["nw"]` -- `D>E>F vs X>Y>E`
        - `logical_detail["bn"]` -- `D>E>F vs X>D>Z`
        - `logical_detail["bw"]` -- `D>E>F vs X>Y>D`
        - `logical_detail["wn"]` -- `D>E>F vs X>F>Z`
        - `logical_detail["wb"]` -- `D>E>F vs F>Y>Z`


**Limit the Database Size:**
Logical Inference has quadratic complexity, and it might be beneficial to limit the database what can be done by the `logical_database` parameter.

```python
import bwsample as bws
agg_dok, direct_dok, direct_detail, logical_dok, logical_detail = bws.count(
    evaluations, logical_database=evaluations[:1])
```

**Update Frequencies:**
The function `bwsample.count` is an update function, i.e. you can provide previous count or resp. frequency data (e.g. `logical_dok`) or start from scratch (e.g. `agg_dok=None`).

```python
import bwsample as bws

evaluations = [...]
direct_dok = {...}
direct_detail = {...}
logical_dok = {...}
logical_detail = {...}
database = [...]

agg_dok, direct_dok, direct_detail, logical_dok, logical_detail = bws.count(
    evaluations, direct_dok=direct_dok, direct_detail=direct_detail,
    logical_dok=logical_dok, logical_detail=logical_detail, logical_database=database)
```

**References:**

- Section 3-4 in: Hamster, U. A. (2021, March 9). Extracting Pairwise Comparisons Data from Best-Worst Scaling Surveys by Logical Inference. [https://doi.org/10.31219/osf.io/qkxej](https://doi.org/10.31219/osf.io/qkxej)


## Ranking
**Input Data:**
The input data is a Dictionary of Keys (DoK) object produced by `bwsample.count`. 

**Call the function:**
The function `bwsample.rank` computes a python index variable with a proposed ordering (`ranked`), and ordered list of example IDs (`ordids`), scores (`scores`) and further information depending on the selected `method`.

```python
import bwsample as bws
ranked, ordids, metrics, scores, info = bws.rank(dok, method='ratio', adjust='quantile')
```

**Available methods:**
Computed from extracted pairs:

- `'ratio'` -- Simple ratios for each pair, and sum ratios for each item.
- `'approx'` -- Chi-Squared based p-value (Hoaglin Approximation) for each pair, and sum 1-pval for each item (Beh et al, 2018)
- `'btl'` -- Bradley-Terry-Luce (BTL) model estimated with MM algorithm (Hunter, 2004).
- `'eigen'` -- Eigenvectors of the reciprocal pairwise comparison matrix (Saaty, 2003).
- `'trans'` -- Estimate transition probability of the next item to be better.

The implementations `ratio`, `pvalue`, `'btl'`, `'eigen'`, and `'trans'` are fully based on sparse matrix operations and `scipy.sparse` algorithms, and avoid accidental conversions to dense matrices.


**References:**
- Hoaglin Approximation for p-values: Beh, E., 2018. Exploring How to Simply Approximate the P-value of a Chi-squared Statistic. AJS 47, 63–75. [https://doi.org/10.17713/ajs.v47i3.757](https://doi.org/10.17713/ajs.v47i3.757)
- Eigenvector solution in: Saaty, T. L. (2003). Decision-making with the AHP: Why is the principal eigenvector nec- essary. European Journal of Operational Research, 145(1), 85–91. [https://doi.org/10.1016/S0377-2217(02)00227-8](https://doi.org/10.1016/S0377-2217(02)00227-8)
- Estimating the BTL model in: Hunter, D. R. (2004). MM algorithms for generalized Bradley-Terry models. The Annals of Statistics, 32(1), 384–406. [https://doi.org/10.1214/aos/1079120141](https://doi.org/10.1214/aos/1079120141)
- MaxDiff score in: Orme, B. (2009). MaxDiff Analysis: Simple Counting, Individual-Level Logit, and HB. [https://sawtoothsoftware.com/uploads/sawtoothsoftware/originals/f89a6537-1cae-4fb5-afad-9d325c2a3143.pdf](https://sawtoothsoftware.com/uploads/sawtoothsoftware/originals/f89a6537-1cae-4fb5-afad-9d325c2a3143.pdf)
- Hamster, U. A. (2021, April 1). Pairwise comparison based ranking and scoring algorithms. [https://doi.org/10.31219/osf.io/ev7fw](https://doi.org/10.31219/osf.io/ev7fw)


## Appendix

### Install a virtual environment
In order to run the Jupyter notebooks or want to work on this project (e.g. unit tests, syntax checks) you should install a Python virtual environment.

```sh
python3.7 -m venv .venv
source .venv/bin/activate
pip install --upgrade pip
pip install -r requirements.txt --no-cache-dir
pip install -r requirements-dev.txt --no-cache-dir
pip install -r requirements-demo.txt --no-cache-dir
```

(If your git repo is stored in a folder with whitespaces, then don't use the subfolder `.venv`. Use an absolute path without whitespaces.)

### Python commands

* Jupyter for the examples: `jupyter lab`
* Check syntax: `flake8 --ignore=F401 --exclude=$(grep -v '^#' .gitignore | xargs | sed -e 's/ /,/g')`
* Run Unit Tests: `pytest`

Publish

```sh
# pandoc README.md --from markdown --to rst -s -o README.rst
python setup.py sdist 
twine check dist/*
twine upload -r pypi dist/*
```

### Clean up 

```sh
find . -type f -name "*.pyc" | xargs rm
find . -type d -name "__pycache__" | xargs rm -r
rm -r .pytest_cache
rm -r .venv
```


### Support
Please [open an issue](https://github.com/satzbeleg/bwsample/issues/new) for support.


### Contributing
Please contribute using [Github Flow](https://guides.github.com/introduction/flow/). Create a branch, add commits, and [open a pull request](https://github.com/satzbeleg/bwsample/compare/).


### Acknowledgements
This work was funded by the Deutsche Forschungsgemeinschaft (DFG, German Research Foundation) - [433249742](https://gepris.dfg.de/gepris/projekt/433249742). Project duration: 2020-2023.

### Citation
Please cite the peer-reviewed JOSS paper [![DOI](https://joss.theoj.org/papers/10.21105/joss.03324/status.svg)](https://doi.org/10.21105/joss.03324) when using this software for any purpose.


CHANGES.md
# 0.7.0 / 2023-02-10

  * Fix dependency problems

# 0.6.9 / 2021-08-13

  * Bump version up for JOSS correction

# 0.6.8 / 2021-08-13

  * Bump version up for JOSS

# 0.6.5 / 2021-04-23

  * Test multiple python versions
  * Installation problems: remove pandoc from setup.py

# 0.6.4 / 2021-04-08

  * PyPi RST format description

# 0.6.3 / 2021-04-01

  * Hoaglin's Approximation for p-value based metric

# 0.6.1 / 2021-03-14

  * repo moved to the organization `github.com/satzbeleg`

# 0.5.5 / 2021-02-23

  * Default settings changed for `transition_simulation`
  * regularize zero-elements in `mle_btl_sparse`

# 0.5.4 / 2021-02-22

  * Scoring by Orme (2009) added to docs
  * Bradley-Terry-Luce (BTL) Model added

# 0.5.2 / 2021-02-22

  * Platt scaling added from scikit-learn
  * sparse averaging for ratio-based and p-value-based ranking

# 0.5.1 / 2021-02-21

  * derive scores by solving an eigenvalue problem

# 0.5.0 / 2021-02-15

  * Code refactoring
  * Main wrapper functions: bws.sampling, bws.counting, bws.ranking

# 0.4.0 / 2021-02-13

  * Identify pairs by logical inference between two BWS sets

# 0.3.0 / 2021-02-12

  * Scaling of just two frequencies based on simple averaging
  * Use p-values of an one-sided Chi-Sequared test as scaled value

# 0.2.0 / 2021-02-12

  * Count pairs from a BWS set
  * Converter to scipy sparse matrix

# 0.1.0 / 2021-02-12

  * Initial Release


bwsample/ranking.py
from typing import List, Dict, Tuple, Optional
from .utils import to_scipy
from .utils import adjustscore
from .utils import minmax
import numpy as np
import scipy.sparse
import scipy.sparse.linalg
import scipy.linalg
import scipy.stats
import warnings


def rank(dok: Dict[Tuple[str, str], int],
         method: Optional[str] = 'ratio',
         adjust: Optional[str] = None,
         **kwargs) -> (np.array, np.array, np.array, dict):
    """Rank items based on pairwise comparison frequencies

    Parameters:
    -----------
    dok : Dict[Tuple[str, str], int]
        Count/Frequency data as Dictionary of Keys (DoK)

    method : Optional[str]
        The procedure to compute ranks and scores.
        - 'ratio'
        - 'pvalue'
        - 'approx'
        - 'btl'
        - 'eigen'
        - 'trans'

    Returns:
    --------
    positions : np.array[uint64]
        The array positions to order/sort the original data by indexing.

    sortedids : np.array[any]
        The reordered item IDs

    metrics : np.array[float]
        The metric for each item ID. Also sorted in descending order.

    scores : np.array[float]
        Scaled or calibrated metrics (Default: The `metrics` if `adjust=None`)

    info : dict
        Further information depending on the selected `method`

    Example:
    --------
        import bwsample as bws
        evaluations = (
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([2, 0, 0, 1], ['A', 'B', 'C', 'D']),
            ([0, 1, 2, 0], ['A', 'B', 'C', 'D']),
            ([0, 1, 0, 2], ['A', 'B', 'C', 'D']),
        )
        agg_dok, _, _, _, _ = bws.count(evaluations)
        positions, sortedids, metrics, info = bws.rank(
            agg_dok, method='ratio', avg='exist', adjust='ordinal')
    """
    # convert to sparse matrix
    cnt, indices = to_scipy(dok)

    # compute the rankings
    if method in ('ratio'):
        positions, sortedids, metrics, info = maximize_ratio(
            cnt, indices, **kwargs)
    elif method in ('pvalue'):
        warnings.warn("Use 'approx' because it's faster.", UserWarning)
        positions, sortedids, metrics, info = maximize_minuspvalue(
            cnt, indices, **kwargs)
    elif method in ('approx', 'hoaglin'):
        positions, sortedids, metrics, info = maximize_hoaglinapprox(
            cnt, indices, **kwargs)
    elif method in ('btl', 'hunter'):
        positions, sortedids, metrics, info = bradley_terry_probability(
            cnt, indices, **kwargs)
    elif method in ('eigen', 'saaty'):
        positions, sortedids, metrics, info = eigenvector_estimation(
            cnt, indices)
    elif method in ('trans'):
        positions, sortedids, metrics, info = transition_simulation(
            cnt, indices, **kwargs)
    else:
        raise Exception(f"method='{method}' not available.")

    # adjust scores
    if adjust is not None:
        cut = np.median(metrics)
        labels = [x >= cut for x in metrics]
        scores = adjustscore(metrics, method=adjust, labels=labels)
    else:
        scores = metrics.copy()

    # done
    return positions, sortedids, metrics, scores, info


def maximize_ratio(cnt: scipy.sparse.csr_matrix,
                   indices: List[str],
                   avg: Optional[str] = 'exist'):
    """Rank items based simple ratios, and calibrate row sums as scores

    Parameters:
    -----------
    cnt : scipy.sparse.dok.dok_matrix
        Quadratic sparse matrix with frequency data

    indices : List[str]
        Identifiers, e.g. UUID4, of each row/column of the `cnt` matrix.

    avg : Optional[str] = 'exists'
        How to compute denominator for averaging.
        - 'all': divide the sum of ratios by the row length
        - 'exist': divide the sum of ratios by the number of ratios in the row

    Returns:
    --------
    positions : np.array[uint64]
        The array positions to order/sort the original data by indexing.

    sortedids : np.array[any]
        The reordered item IDs

    metrics : np.array[float]
        The metric for each item ID. Also sorted in descending order.

    info : dict
        Further information depending on the selected `method`

    Example:
    --------
        import bwsample as bws
        evaluations = (
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([2, 0, 0, 1], ['A', 'B', 'C', 'D']),
            ([0, 1, 2, 0], ['A', 'B', 'C', 'D']),
            ([0, 1, 0, 2], ['A', 'B', 'C', 'D']),
        )
        agg_dok, _, _, _, _ = bws.count(evaluations)
        positions, sortedids, metrics, info = bws.rank(
            agg_dok, method='ratio', avg='exist')
    """
    # compute ratios
    cnt = cnt.tocsr()
    ratios = cnt + cnt.T
    ratios.data = 1.0 / ratios.data
    ratios = ratios.multiply(cnt)

    # sum rows in DoK matrix
    metrics = np.array(ratios.sum(axis=1).flatten())[0]
    # averaging
    if avg == 'all':
        metrics /= len(metrics)
    elif avg == 'exist':
        ridx, _ = (ratios + ratios.T).nonzero()  # ensure actual 0s are counted
        for i, c in zip(*np.unique(ridx, return_counts=True)):
            metrics[i] /= c

    # sort, larger row sums are better
    positions = np.argsort(-metrics)  # maximize
    sortedids = np.array(indices)[positions]
    metrics = metrics[positions]

    # informations
    info = {}

    # done
    return positions, sortedids, metrics, info


def maximize_minuspvalue(cnt: scipy.sparse.csr_matrix,
                         indices: List[str],
                         avg: Optional[str] = 'exist'):
    """Rank based on p-values of a Chi-Squard tests between reciprocal pairs,
        and calibrate row sums as scores

    Parameters:
    -----------
    cnt : scipy.sparse.dok.dok_matrix
        Quadratic sparse matrix with frequency data

    indices : List[str]
        Identifiers, e.g. UUID4, of each row/column of the `cnt` matrix.

    avg : Optional[str]
        How to compute denominator for averaging.
        - 'all': divide the sum of ratios by the row length
        - 'exist': divide the sum of ratios by the number of ratios in the row

    Returns:
    --------
    positions : np.array[uint64]
        The array positions to order/sort the original data by indexing.

    sortedids : np.array[any]
        The reordered item IDs

    metrics : np.array[float]
        The metric for each item ID. Also sorted in descending order.

    info : dict
        Further information depending on the selected `method`, e.g.
        - "P": The matrix with the `1-p`-values

    Example:
    --------
        import bwsample as bws
        evaluations = (
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([2, 0, 0, 1], ['A', 'B', 'C', 'D']),
            ([0, 1, 2, 0], ['A', 'B', 'C', 'D']),
            ([0, 1, 0, 2], ['A', 'B', 'C', 'D']),
        )
        agg_dok, _, _, _, _ = bws.count(evaluations)
        positions, sortedids, metrics, info = bws.rank(
            agg_dok, method='pvalue', avg='exist')
    """
    # compute p-values for Nij>Nji or 1
    n, _ = cnt.shape
    P = scipy.sparse.dok_matrix((n, n), dtype=np.float64)
    for i in range(n):
        for j in range(n):
            if i > j:
                f1, f2 = cnt[i, j], cnt[j, i]
                fe = (f1 + f2) / 2  # E: Expected Frequency
                if fe > 0 and f1 != f2:
                    _, pval = scipy.stats.chisquare(
                        f_obs=[f1, f2], f_exp=[fe, fe], ddof=0)
                    if f1 > f2:
                        P[i, j] = 1 - pval
                    else:
                        P[j, i] = 1 - pval

    # sum rows in DoK matrix
    metrics = np.array(P.sum(axis=1).flatten())[0]
    # averaging
    if avg == 'all':
        metrics /= len(metrics)
    elif avg == 'exist':
        ridx, _ = (P + P.T).nonzero()  # ensure actual 0s are counted
        for i, c in zip(*np.unique(ridx, return_counts=True)):
            metrics[i] /= c

    # sort, larger row sums are better
    positions = np.argsort(-metrics)  # minimize P, maximize 1-P
    sortedids = np.array(indices)[positions]
    metrics = metrics[positions]

    # informations
    info = {}
    info["P"] = P

    # done
    return positions, sortedids, metrics, info


def maximize_hoaglinapprox(cnt: scipy.sparse.csr_matrix,
                           indices: List[str],
                           avg: Optional[str] = 'exist'):
    """Rank based on p-values computed with the Hoaglin Approximation of DoF=0

    Parameters:
    -----------
    cnt : scipy.sparse.dok.dok_matrix
        Quadratic sparse matrix with frequency data

    indices : List[str]
        Identifiers, e.g. UUID4, of each row/column of the `cnt` matrix.

    avg : Optional[str]
        How to compute denominator for averaging.
        - 'all': divide the sum of ratios by the row length
        - 'exist': divide the sum of ratios by the number of ratios in the row

    Returns:
    --------
    positions : np.array[uint64]
        The array positions to order/sort the original data by indexing.

    sortedids : np.array[any]
        The reordered item IDs

    metrics : np.array[float]
        The metric for each item ID. Also sorted in descending order.

    info : dict
        Further information depending on the selected `method`, e.g.
        - "P": The matrix with the `1-p`-values

    Example:
    --------
        import bwsample as bws
        evaluations = (
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([2, 0, 0, 1], ['A', 'B', 'C', 'D']),
            ([0, 1, 2, 0], ['A', 'B', 'C', 'D']),
            ([0, 1, 0, 2], ['A', 'B', 'C', 'D']),
        )
        agg_dok, _, _, _, _ = bws.count(evaluations)
        positions, sortedids, metrics, info = bws.rank(
            agg_dok, method='approx', avg='exist')
    """
    # compute Expected E
    cnt = cnt.tocsr()
    E = cnt + cnt.T
    E.data = E.data / 2.0
    # compute X^2
    X2 = cnt - E
    X2.data = (X2.data)**2
    E.data = 1.0 / E.data
    X2 = X2.multiply(E)
    # compute Hoaglin's Approximation for DoF=0
    P = np.sqrt(X2)
    P.data = np.power(0.1, (P.data + 1.37266) / 2.13161)
    # P.data = P.data * 4.405087805849058
    P.data = np.maximum(0.0, np.minimum(1.0, P.data))
    # only if Nij>Nji
    P = P.multiply(cnt > cnt.T)
    # Q = 1-P
    P.data = 1. - P.data

    # sum rows in DoK matrix
    metrics = np.array(P.sum(axis=1).flatten())[0]
    # averaging
    if avg == 'all':
        metrics /= len(metrics)
    elif avg == 'exist':
        ridx, _ = (P + P.T).nonzero()  # ensure actual 0s are counted
        for i, c in zip(*np.unique(ridx, return_counts=True)):
            metrics[i] /= c

    # sort, larger row sums are better
    positions = np.argsort(-metrics)  # minimize P, maximize 1-P
    sortedids = np.array(indices)[positions]
    metrics = metrics[positions]

    # informations
    info = {}
    info["P"] = P

    # done
    return positions, sortedids, metrics, info


def eigenvector_estimation(cnt: scipy.sparse.csr_matrix,
                           indices: List[str]):
    """Compute the eigenvectors of the pairwise comparison matrix, and
        calibrate eigenvectors as scores.

    Parameters:
    -----------
    cnt : scipy.sparse.dok.dok_matrix
        Quadratic sparse matrix with frequency data

    indices : List[str]
        Identifiers, e.g. UUID4, of each row/column of the `cnt` matrix.

    Returns:
    --------
    positions : np.array[uint64]
        The array positions to order/sort the original data by indexing.

    sortedids : np.array[any]
        The reordered item IDs

    metrics : np.array[float]
        The metric for each item ID. Also sorted in descending order.

    info : dict
        Further information depending on the selected `method`, e.g.
        - "eigval": Estimated eigenvalue
        - "eigenvec": Estimated eigenvector

    Example:
    --------
        import bwsample as bws
        evaluations = (
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([2, 0, 0, 1], ['A', 'B', 'C', 'D']),
            ([0, 1, 2, 0], ['A', 'B', 'C', 'D']),
            ([0, 1, 0, 2], ['A', 'B', 'C', 'D']),
        )
        agg_dok, _, _, _, _ = bws.count(evaluations)
        positions, sortedids, metrics, info = bws.rank(
            agg_dok, method='eigen')

    References:
    -----------
    Saaty, T.L., 2003. Decision-making with the AHP: Why is the principal
        eigenvector necessary. European Journal of Operational Research 145,
        85–91. https://doi.org/10.1016/S0377-2217(02)00227-8
    """
    # set diagonals to 1
    n = cnt.shape[0]
    cnt = cnt.tolil()
    for i in range(n):
        cnt[i, i] = 1

    # Compute a sparse "positive reciprocal near consistent pairwise
    #   comparison matrix". Avoid accidental conversion into dense matrix
    #   by manipulating the value/data vector of the transposed sp matrix.
    cnt = cnt.tocsr()
    cntT = cnt.T
    cntT.data = 1.0 / cntT.data
    ratios = cnt.multiply(cntT)

    # compute eigenvectors as scores
    eigval, eigenvec = scipy.sparse.linalg.eigs(ratios, k=1)
    metrics = np.abs(np.real(eigenvec[:, 0]))

    # sort, larger row sums are better
    positions = np.argsort(-metrics)  # maximize
    sortedids = np.array(indices)[positions]
    metrics = metrics[positions]

    # informations
    info = {}
    info["eigval"] = eigval
    info["eigenvec"] = eigenvec

    # done
    return positions, sortedids, metrics, info


def transition_simulation(cnt: scipy.sparse.dok.dok_matrix,
                          indices: List[str],
                          n_rounds: Optional[int] = 2):
    """Estimate transition matrix of item_i>item_j, simulate the item
        probabilities that are calibrated to scores.

    Parameters:
    -----------
    cnt : scipy.sparse.dok.dok_matrix
        Quadratic sparse matrix with frequency data

    indices : List[str]
        Identifiers, e.g. UUID4, of each row/column of the `cnt` matrix.

    n_rounds: Optional[int] = 2
        Number of steps/rounds to simulate

    Returns:
    --------
    positions : np.array[uint64]
        The array positions to order/sort the original data by indexing.

    sortedids : np.array[any]
        The reordered item IDs

    metrics : np.array[float]
        The metric for each item ID. Also sorted in descending order.

    info : dict
        Further information depending on the selected `method`, e.g.
        - "sim": The predicted/simulated item probability
        - "transmat: The estimated transition probability matrix.

    Example:
    --------
        import bwsample as bws
        evaluations = (
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([2, 0, 0, 1], ['A', 'B', 'C', 'D']),
            ([0, 1, 2, 0], ['A', 'B', 'C', 'D']),
            ([0, 1, 0, 2], ['A', 'B', 'C', 'D']),
        )
        agg_dok, _, _, _, _ = bws.count(evaluations)
        positions, sortedids, metrics, info = bws.rank(
            agg_dok, method='trans', n_rounds=3)
    """
    n = cnt.shape[0]

    # create generator matrix
    genmat = cnt.T.tolil()
    rowsum = genmat.sum(axis=1)
    for i in range(n):
        genmat[i, i] = -rowsum[i]  # set diagonals to -sum(row)
        genmat[i, :] /= max(1, rowsum[i])

    # compute transition matrix
    transmat = scipy.sparse.linalg.expm(genmat.tocsc())

    # simulation: transition from an item1 to the next item2
    #   that is most likely "item2 > item1"
    x = np.ones(n) / n
    for i in range(n_rounds):
        x = x * transmat

    # sort, larger state probabilities are better
    positions = np.argsort(-x)  # maximize
    sortedids = np.array(indices)[positions]
    metrics = x[positions]

    # informations
    info = {}
    info["sim"] = x
    info["transmat"] = transmat

    # done
    return positions, sortedids, metrics, info


def mle_btl_sparse(cnt: scipy.sparse.csr_matrix,
                   x0: Optional[np.array] = None,
                   max_iter: Optional[int] = 50,
                   tol: Optional[float] = 1e-5) -> (np.array, bool):
    """MLE by Hunter (2004, p.386-387)

    Parameters:
    -----------
    cnt : scipy.sparse.dok.dok_matrix
        Quadratic sparse matrix with frequency data

    x0 : np.array
        Initial values.

    max_iter : int
        maximum number of iterations

    tol : float
        termination criteria

    Returns:
    --------
    gamma : np.array
        Estimated gamma parameters. SUM[gammas]=1. The estimated
          parameters can be used as scores

    flag : bool
        True if solution was found within `max_iter` optimization steps.
        False if not.

    References:
    -----------
    Hunter, D.R., 2004. MM algorithms for generalized Bradley-Terry models. The
      Annals of Statistics 32, 384–406. https://doi.org/10.1214/aos/1079120141
    """
    # ensure CSR format
    cnt = cnt.tocsr()
    m = cnt.shape[0]

    # set initial values
    if x0 is None:
        x = np.ones(m) / m
    else:
        x = np.array(x0)
        x = x / x.sum()

    # rowsum, i.e. the number of wins `W_i`
    rowsum = cnt.sum(axis=1)

    # count zero rows
    zrow = m - (cnt > 0).sum(axis=1)

    # add `Nij + Nji`
    cntij = (cnt + cnt.T)
    # pseudo inverse `[(Nij + Nji) / (yi + yj)]^{-1}`
    cntij.data = 1 / cntij.data

    # copy sparse structure for `yi + yj`
    gamij = cntij.copy()
    ridx, cidx = gamij.nonzero()

    for k in range(max_iter):
        # assign new weights
        gamij[(ridx, cidx)] = x[ridx] + x[cidx]

        # don't forget to use the inverse:
        tmp = gamij.multiply(cntij)

        # sum up rows, and elementwise multiply
        gamk = np.multiply(rowsum, zrow + tmp.sum(axis=1))

        # normalize to `gam_i^(k)`
        x1 = gamk / gamk.sum(axis=0)

        # abort
        if np.linalg.norm(x1 - x, ord=np.inf) < tol:
            return np.array(x1.flatten())[0], True

        # update
        x = x1

    # last result
    return np.array(x1.flatten())[0], False


def bradley_terry_probability(cnt: scipy.sparse.csr_matrix,
                              indices: List[str],
                              prefit: Optional[bool] = True,
                              max_iter: Optional[int] = 50,
                              tol: Optional[float] = 1e-5):
    """Bradley-Terry-Luce (BTL) probability model for pairwise comparisons

    Parameters:
    -----------
    cnt : scipy.sparse.dok.dok_matrix
        Quadratic sparse matrix with frequency data

    indices : List[str]
        Identifiers, e.g. UUID4, of each row/column of the `cnt` matrix.

    prefit : bool
        flag to prefit parameters with 'ratio' method
        (see `ranking_maximize_ratios`)

    max_iter : int  (see `mle_btl_sparse`)
        maximum number of iterations

    tol : float  (see `mle_btl_sparse`)
        termination criteria

    Returns:
    --------
    positions : np.array[uint64]
        The array positions to order/sort the original data by indexing.

    sortedids : np.array[any]
        The reordered item IDs

    metrics : np.array[float]
        The metric for each item ID. Also sorted in descending order.

    info : dict
        Further information depending on the selected `method`, e.g.
        - "weights": The estimated MLE parameters that can be used for scoring

    Example:
    --------
        import bwsample as bws
        evaluations = (
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([1, 0, 0, 2], ['A', 'B', 'C', 'D']),
            ([2, 0, 0, 1], ['A', 'B', 'C', 'D']),
            ([0, 1, 2, 0], ['A', 'B', 'C', 'D']),
            ([0, 1, 0, 2], ['A', 'B', 'C', 'D']),
        )
        agg_dok, _, _, _, _ = bws.count(evaluations)
        positions, sortedids, metrics, info = bws.rank(
            agg_dok, method='btl', prefit=True, max_iter=100, tol=1e-5)
    """
    cnt = cnt.tocsr()
    x0 = None
    if prefit:
        ratios = cnt + cnt.T
        ratios.data = 1.0 / ratios.data
        ratios = ratios.multiply(cnt)
        x0 = np.array(ratios.sum(axis=1).flatten())[0]
        x0 = minmax(x0)

    # estimate Bradley-Terry-Luce model parameters as metric
    x, flag = mle_btl_sparse(cnt, x0=x0, max_iter=max_iter, tol=tol)

    # sort, larger state probabilities are better
    positions = np.argsort(-x)  # maximize
    sortedids = np.array(indices)[positions]
    metrics = x[positions]

    # informations
    info = {}
    info["weights"] = x

    # done
    return positions, sortedids, metrics, info
